src/main/java/org/apache/commons/io/input/TeeReader.java

```diff
@@ -0,0 +1,180 @@
+    private final Writer branch;
+
+    /**
+     * Flag for closing also the associated writer when this
+     * reader is closed.
+     */
+    private final boolean closeBranch;
+
+    /**
+     * Creates a TeeReader that proxies the given {@link Reader}
+     * and copies all read characters to the given {@link Writer}. The given
+     * writer will not be closed when this reader gets closed.
+     *
+     * @param input reader to be proxied
+     * @param branch writer that will receive a copy of all characters read
+     */
+    public TeeReader(final Reader input, final Writer branch) {
+        this(input, branch, false);
+    }
+
+    /**
+     * Creates a TeeReader that proxies the given {@link Reader}
+     * and copies all read characters to the given {@link Writer}. The given
+     * writer will be closed when this reader gets closed if the
+     * closeBranch parameter is {@code true}.
+     *
+     * @param input reader to be proxied
+     * @param branch writer that will receive a copy of all characters read
+     * @param closeBranch flag for closing also the writer when this
+     *                    reader is closed
+     */
+    public TeeReader(
+            final Reader input, final Writer branch, final boolean closeBranch) {
+        super(input);
+        this.branch = branch;
+        this.closeBranch = closeBranch;
+    }
+
+    /**
+     * Closes the proxied reader and, if so configured, the associated
+     * writer. An exception thrown from the reader will not prevent
+     * closing of the writer.
+     *
+     * @throws IOException if either the reader or writer could not be closed
+     */
+    @Override
+    public void close() throws IOException {
+        try {
+            super.close();
+        } finally {
+            if (closeBranch) {
+                branch.close();
+            }
+        }
+    }
+
+    /**
+     * Reads a single chracter from the proxied reader and writes it to
+     * the associated writer.
+     *
+     * @return next character from the reader, or -1 if the reader has ended
+     * @throws IOException if the reader could not be read (or written)
+     */
+    @Override
+    public int read() throws IOException {
+        final int ch = super.read();
+        if (ch != EOF) {
+            branch.write(ch);
+        }
+        return ch;
+    }
+
+    /**
+     * Reads characters from the proxied reader and writes the read characters
+     * to the associated writer.
+     *
+     * @param chr character buffer
+     * @param st start offset within the buffer
+     * @param end maximum number of characters to read
+     * @return number of characters read, or -1 if the reader has ended
+     * @throws IOException if the reader could not be read (or written)
+     */
+    @Override
+    public int read(final char[] chr, final int st, final int end) throws IOException {
+        final int n = super.read(chr, st, end);
+        if (n != -1) {
+            branch.write(chr, st, n);
```

Comment:
Use EOF in the test.

Reply:
Fixed, also in TeeInputStream where I copied the text from.